### PR TITLE
[Marketplace Contribution] - EXPANDR-7038 - Azure Resource Graph

### PR DESCRIPTION
**Pull request by BigEasyJ**
## Contributing to Cortex XSOAR Content
Make sure to register your contribution by filling the [contribution registration form](https://forms.gle/XDfxU4E61ZwEESSMA)

**The Pull Request will be reviewed only after the contribution registration form is filled.**

## Open Questions

- Can this pack initial only be for Xpanse and XSIAM marketplaces?

## Remaining Items:
- [ ] WIP - **Creating a video for Demo**
- [ ] Doc review

## Status
- [x] In Progress
- [ ] Ready
- [ ] In Hold - (Reason for hold)

## Related Issues
[EXPANDR-7038](https://jira-dc.paloaltonetworks.com/browse/EXPANDR-7038)
[CIAC-9474](https://jira-dc.paloaltonetworks.com/browse/CIAC-9474)

## Description
This is a new pack to help support enrichment needs for Azure resources. It is difficult and time consuming today to loop through existing Azure APIs for resource information and Azure Resource Graph can do that in a fraction of time and pull together much information in a single query.

## Must have
- [x] Tests
- [x] Documentation

### CODE REVIEW

**Comment by BigEasyJ** [comment]
@ShirleyDenkberg please doc review when you get a chance.

**Comment by ShirleyDenkberg** [comment]
@shmuel44 Doc review completed.

**Comment by RotemAmit** [comment]
Hi @BigEasyJ,
Since this is a new pack for an Azure product, it will be xsoar supported.
For this pack to match our standard of the other Azure packs, we opened a [ticket](https://jira-dc.paloaltonetworks.com/browse/CIAC-9474) to create a full design. The ticket will get a high priority. 
As soon as the design is ready we will continue the review process.
Thank you for your understanding and contribution!

**Review by JasBeilin — COMMENTED** [review]
Great, Please check my comments :)

**Comment by JasBeilin** [comment]
Hi @BigEasyJ, Haven't heard from you in a while, Let me know if you need any help in moving forward here.

File: Packs/AzureResourceGraph/Integrations/AzureResourceGraph/AzureResourceGraph.yml [hunk 1]
```diff
@@ -0,0 +1,121 @@
...
+    - defaultValue: '50'
+      description: The maximum number of operations to return (Default is 50).
+      name: limit
+    outputs:
+    - contextPath: AzureResourceGraph.Operations
+      description: A list of available Azure Resource Graph operations permissions and descriptions.
+      type: String
+    description: Gets all Azure Resource Graph operations permissions and descriptions.
+  - name: azure-rg-query
+    arguments:
```
[the remark this answers is not in the document]
Comment (BigEasyJ) [reply Packs/AzureResourceGraph/Integrations/AzureResourceGraph/AzureResourceGraph.yml:108]: I'm not sure we need:

management_groups
subscriptions

File: Packs/AzureResourceGraph/Integrations/AzureResourceGraph/AzureResourceGraph.py [hunk 2]
```diff
@@ -0,0 +1,248 @@
...
+    }
+
+    '''EXECUTION'''
+    command = demisto.command()
+    LOG(f'Command being called is {command}')
+
+    try:
+        # Initial setup
+        if not subscription_id:
+            return_error('A subscription ID must be provided.')
```
Comment (JasBeilin) [inline Packs/AzureResourceGraph/Integrations/AzureResourceGraph/AzureResourceGraph.py:222]: Its better practice to take this logic outside of the main.

**Review by JasBeilin — COMMENTED** [review]
Hi, thank you for the effort here!

please note my comments :)

File: Packs/AzureResourceGraph/README.md [hunk 3]
```diff
@@ -0,0 +1,27 @@
...
+## What does this pack do?
+
+You can use Azure Resource Graph queries to:
+
+- Query resources with complex filtering, grouping, and sorting by resource properties.
+- Explore resources iteratively based on governance requirements.
+- Assess the effect of applying policies in a vast cloud environment.
+- Query changes made to resource properties.
+
+To access Azure Log Analytics, you need Cortex XSOAR authorization. For details, see https://xsoar.pan.dev/docs/reference/integrations/azure-log-analytics.
```
Comment (michal-dagan) [inline Packs/AzureResourceGraph/README.md:12]: please remove

File: Packs/AzureResourceGraph/Integrations/AzureResourceGraph/AzureResourceGraph_description.md [hunk 4]
```diff
@@ -0,0 +1,25 @@
+Azure Resource Graph is an Azure service designed to extend Azure Resource Management by providing efficient and performant resource exploration with the ability to query at scale across a given set of resources.
+
+Full documentation for this integration is available in the [reference docs](https://xsoar.pan.dev/docs/reference/integrations/azure-resource-graph).
+
+
+## Authorize Cortex XSOAR for Azure Log Analytics - Self-Deployed Configuration
+To use a self-configured Azure application, you need to add a new Azure App Registration in the Azure Portal. To add the registration, go to the [Microsoft article](https://docs.microsoft.com/en-us/azure/active-directory/develop/quickstart-register-app).
+
+### Required permissions
```
Comment (michal-dagan) [inline Packs/AzureResourceGraph/Integrations/AzureResourceGraph/AzureResourceGraph_description.md:9]: please add the required permissions:)

File: Packs/AzureResourceGraph/Integrations/AzureResourceGraph/README.md [hunk 5]
```diff
@@ -0,0 +1,182 @@
+Azure Resource Graph is an Azure service designed to extend Azure Resource Management by providing efficient and performant resource exploration with the ability to query at scale across a given set of resources.
+
+## Authorize Cortex XSOAR for Azure Resource Graph (Self-Deployed Configuration)
+
+To use a self-configured Azure application, you need to add a new Azure App Registration in the Azure Portal. To add the registration, see the [Microsoft article](https://docs.microsoft.com/en-us/azure/active-directory/develop/quickstart-register-app).
+
+## Authentication
+For more details about the authentication used in this integration, see [Microsoft Integrations - Authentication](https://xsoar.pan.dev/docs/reference/articles/microsoft-integrations---authentication).
+
+- After authorizing the XSOAR App or the Self-Deployed Application, you will get an ID, Token, and Key, which should be inserted in the integration instance configuration's corresponding fields. After giving consent, the application has to have a role assigned so it can access the relevant resources per subscription.
```
Comment (michal-dagan) [inline Packs/AzureResourceGraph/Integrations/AzureResourceGraph/README.md:10]: XSOAR App does not exist for this integration, please leave only the reference to Self-Deployed

File: Packs/AzureResourceGraph/Integrations/AzureResourceGraph/AzureResourceGraph_description.md (hunk 4 above)
Comment (BigEasyJ) [reply Packs/AzureResourceGraph/Integrations/AzureResourceGraph/AzureResourceGraph_description.md:9]: It says `user_impersenation` however I did not need it to perform REST calls and it is not very user friendly to use if it it was required.

File: Packs/AzureResourceGraph/Integrations/AzureResourceGraph/AzureResourceGraph_description.md [hunk 6]
```diff
@@ -0,0 +1,25 @@
+Azure Resource Graph is an Azure service designed to extend Azure Resource Management by providing efficient and performant resource exploration with the ability to query at scale across a given set of resources.
+
+Full documentation for this integration is available in the [reference docs](https://xsoar.pan.dev/docs/reference/integrations/azure-resource-graph).
+
+
+## Authorize Cortex XSOAR for Azure Log Analytics - Self-Deployed Configuration
+To use a self-configured Azure application, you need to add a new Azure App Registration in the Azure Portal. To add the registration, go to the [Microsoft article](https://docs.microsoft.com/en-us/azure/active-directory/develop/quickstart-register-app).
```
[the remark this answers is not in the document]
Comment (BigEasyJ) [reply Packs/AzureResourceGraph/Integrations/AzureResourceGraph/AzureResourceGraph_description.md:7]: Yes

File: Packs/AzureResourceGraph/Integrations/AzureResourceGraph/AzureResourceGraph.yml (hunk 1 above)
[the remark this answers is not in the document]
Comment (BigEasyJ) [reply Packs/AzureResourceGraph/Integrations/AzureResourceGraph/AzureResourceGraph.yml:116]: Updated in b8ee973

File: Packs/AzureResourceGraph/Integrations/AzureResourceGraph/README.md (hunk 5 above)
Comment (BigEasyJ) [reply Packs/AzureResourceGraph/Integrations/AzureResourceGraph/README.md]: Updated in dff3feb

File: Packs/AzureResourceGraph/Integrations/AzureResourceGraph/AzureResourceGraph_description.md (hunk 4 above)
Comment (BigEasyJ) [reply Packs/AzureResourceGraph/Integrations/AzureResourceGraph/AzureResourceGraph_description.md]: It says `user_impersenation` however I did not need it to perform REST calls and it is not very user friendly to use if it it was required.

File: Packs/AzureResourceGraph/README.md (hunk 3 above)
Comment (BigEasyJ) [reply Packs/AzureResourceGraph/README.md]: Updated in de1de8

File: Packs/AzureResourceGraph/Integrations/AzureResourceGraph/AzureResourceGraph.py (hunk 2 above)
Comment (BigEasyJ) [reply Packs/AzureResourceGraph/Integrations/AzureResourceGraph/AzureResourceGraph.py:254]: Updated in dff3feb

File: Packs/AzureResourceGraph/Integrations/AzureResourceGraph/AzureResourceGraph.py [hunk 7]
```diff
@@ -0,0 +1,248 @@
...
+    subscription_id = args.get('subscription_id') or params.get(
+        'cred_subscription_id', {}).get('password') or params.get('subscription_id')
+    proxy: bool = params.get('proxy', False)
+    self_deployed: bool = params.get('self_deployed', False)
+    if not self_deployed and not enc_key:
+        raise DemistoException('Key must be provided. For further information see '
+                               'https://xsoar.pan.dev/docs/reference/articles/microsoft-integrations---authentication')
+    elif not enc_key and not (certificate_thumbprint and private_key):
+        raise DemistoException('Key or Certificate Thumbprint and Private Key must be providedFor further information see '
+                               'https://xsoar.pan.dev/docs/reference/articles/microsoft-integrations---authentication')
```
[the remark this answers is not in the document]
Comment (BigEasyJ) [reply Packs/AzureResourceGraph/Integrations/AzureResourceGraph/AzureResourceGraph.py]: Updated in dff3feb

File: Packs/AzureResourceGraph/Integrations/AzureResourceGraph/AzureResourceGraph.py [hunk 8]
```diff
@@ -0,0 +1,248 @@
...
+    ok_codes = (200, 201, 202, 204)
```
[the remark this answers is not in the document]
Comment (BigEasyJ) [reply Packs/AzureResourceGraph/Integrations/AzureResourceGraph/AzureResourceGraph.py:233]: Updated in dff3feb

File: Packs/AzureResourceGraph/Integrations/AzureResourceGraph/AzureResourceGraph.py [hunk 9]
```diff
@@ -0,0 +1,248 @@
...
+        Return a list of objects from the response according to the page and limit per page.
+    """
+    if page_size > MAX_PAGE_SIZE:
+        page_size = MAX_PAGE_SIZE
+    
+    starting_index = (page_number - 1) * page_size
+    ending_index = starting_index + page_size
+    return response[starting_index:ending_index]
+
+
```
[the remark this answers is not in the document]
Comment (BigEasyJ) [reply Packs/AzureResourceGraph/Integrations/AzureResourceGraph/AzureResourceGraph.py:199]: Updated in dff3feb

File: Packs/AzureResourceGraph/Integrations/AzureResourceGraph/AzureResourceGraph.py [hunk 10]
```diff
@@ -0,0 +1,248 @@
...
+            return_results(commands_without_args[command](client))
+
+        elif command in commands_with_args:
+            return_results(commands_with_args[command](client, args))
+
+        elif command in commands_with_args_and_params:
+            return_results(commands_with_args_and_params[command](client, args, params))
+
+    except Exception as e:
+        return_error(e)
```
[the remark this answers is not in the document]
Comment (BigEasyJ) [reply Packs/AzureResourceGraph/Integrations/AzureResourceGraph/AzureResourceGraph.py]: Updated in dff3feb

File: Packs/AzureResourceGraph/Integrations/AzureResourceGraph/AzureResourceGraph.yml (hunk 1 above)
[the remark this answers is not in the document]
Comment (BigEasyJ) [reply Packs/AzureResourceGraph/Integrations/AzureResourceGraph/AzureResourceGraph.yml:116]: Updated in b8ee9

File: Packs/AzureResourceGraph/Integrations/AzureResourceGraph/AzureResourceGraph.py [hunk 11]
```diff
@@ -0,0 +1,281 @@
...
+        'test-module': test_module
+    }
+
+    commands_with_args: Dict[Any, Any] = {
+        'azure-rg-query': query_resources_command,
+        'azure-rg-list-operations': list_operations_command
+    }
+
+    commands_with_args_and_params: Dict[Any, Any] = {
+    }
```
Comment (JasBeilin) [inline Packs/AzureResourceGraph/Integrations/AzureResourceGraph/AzureResourceGraph.py:245]:
```suggestion
```

Removing as we usually do not pass the params but saves in the client what might be needed.

File: Packs/AzureResourceGraph/Integrations/AzureResourceGraph/AzureResourceGraph.py [hunk 12]
```diff
@@ -0,0 +1,281 @@
...
+            return_results(reset_auth())
+
+        elif command in commands_without_args:
+            return_results(commands_without_args[command](client))
+
+        elif command in commands_with_args:
+            return_results(commands_with_args[command](client, args))
+
+        elif command in commands_with_args_and_params:
+            return_results(commands_with_args_and_params[command](client, args, params))
```
Comment (JasBeilin) [inline Packs/AzureResourceGraph/Integrations/AzureResourceGraph/AzureResourceGraph.py:273]:
```suggestion
```

File: Packs/AzureResourceGraph/Integrations/AzureResourceGraph/AzureResourceGraph.yml [hunk 13]
```diff
@@ -0,0 +1,142 @@
...
+name: Azure Resource Graph
+script:
+  commands:
+  - name: azure-rg-auth-reset
+    arguments: []
+    description: Run this command if for some reason you need to rerun the authentication process.
+  - name: azure-rg-list-operations
+    arguments:
+    - description: 'The maximum number of operations to return. NOTE: Do not use with "page" and "page_size".'
+      name: limit
```
Comment (JasBeilin) [inline Packs/AzureResourceGraph/Integrations/AzureResourceGraph/AzureResourceGraph.yml:105]:
```suggestion
      name: limit
      default: 50
```

File: Packs/AzureResourceGraph/Integrations/AzureResourceGraph/AzureResourceGraph.yml [hunk 14]
```diff
@@ -0,0 +1,142 @@
...
+    - description: The maximum number of operations to return for page.
+      name: page_size
```
Comment (JasBeilin) [inline Packs/AzureResourceGraph/Integrations/AzureResourceGraph/AzureResourceGraph.yml:107]:
```suggestion
      name: page_size
      default: 50
```

File: Packs/AzureResourceGraph/Integrations/AzureResourceGraph/AzureResourceGraph.yml [hunk 15]
```diff
@@ -0,0 +1,142 @@
...
+    - description: The page number to return.
+      name: page
```
Comment (JasBeilin) [inline Packs/AzureResourceGraph/Integrations/AzureResourceGraph/AzureResourceGraph.yml:109]:
```suggestion
      name: page
      default: 1
```

File: Packs/AzureResourceGraph/Integrations/AzureResourceGraph/AzureResourceGraph.yml [hunk 16]
```diff
@@ -0,0 +1,142 @@
...
+  type: 9
+  required: false
+  hiddenusername: true
+- display: ""
+  displaypassword: Certificate Thumbprint
+  name: cred_certificate_thumbprint
+  type: 9
+  required: false
+  hiddenusername: true
+  additionalinfo: Used for certificate authentication. As appears in the "Certificates & secrets" page of the app.
```
Comment (JasBeilin) [inline Packs/AzureResourceGraph/Integrations/AzureResourceGraph/AzureResourceGraph.yml:64]:
```suggestion
  additionalinfo: Used for certificate authentication. As appears in the "Certificates & secrets" page of the app.
  section: connect
```

File: Packs/AzureResourceGraph/Integrations/AzureResourceGraph/AzureResourceGraph.yml [hunk 17]
```diff
@@ -0,0 +1,142 @@
...
+- display: Private Key
+  name: private_key
+  type: 14
+  required: false
+  additionalinfo: Used for certificate authentication. The private key of the registered certificate.
```
Comment (JasBeilin) [inline Packs/AzureResourceGraph/Integrations/AzureResourceGraph/AzureResourceGraph.yml:69]:
```suggestion
  additionalinfo: Used for certificate authentication. The private key of the registered certificate.
  section: connect
```

File: Packs/AzureResourceGraph/Integrations/AzureResourceGraph/AzureResourceGraph_description.md [hunk 18]
```diff
@@ -0,0 +1,23 @@
...
+### Client Credentials Flow
+
+---
+Follow these steps for client-credentials configuration.
+
+1. In the instance configuration under **Advanced Settings**, select the **Use a self-deployed Azure application** checkbox.
+2. Enter your Client ID in the **ID / Client ID** parameter. 
+3. Enter your Client Secret in the **Key / Client Secret** parameter.
+4. Enter your Tenant ID in the **Tenant ID** parameter.
+5. Click **Test** to validate the URLs, token, and connection.
```
Comment (JasBeilin) [inline Packs/AzureResourceGraph/Integrations/AzureResourceGraph/AzureResourceGraph_description.md:23]:
```suggestion

There are two application authentication methods available.
Follow your prefered method's guide on how to use the admin consnt flow in order to recieve your authentication information:

* [Cortex XSOAR Application](https://xsoar.pan.dev/docs/reference/articles/microsoft-integrations---authentication#cortex-xsoar-application)
* [Self-Deployed Application](https://xsoar.pan.dev/docs/reference/articles/microsoft-integrations---authentication#self-deployed-application)

For more details about the authentication used in this integration, see [Microsoft Integrations - Authentication](https://xsoar.pan.dev/docs/reference/articles/microsoft-integrations---authentication).



```

File: Packs/AzureResourceGraph/Integrations/AzureResourceGraph/README.md [hunk 19]
```diff
@@ -0,0 +1,182 @@
+Azure Resource Graph is an Azure service designed to extend Azure Resource Management by providing efficient and performant resource exploration with the ability to query at scale across a given set of resources.
+
+## Authorize Cortex XSOAR for Azure Resource Graph (Self-Deployed Configuration)
+
+To use a self-configured Azure application, you need to add a new Azure App Registration in the Azure Portal. To add the registration, see the [Microsoft article](https://docs.microsoft.com/en-us/azure/active-directory/develop/quickstart-register-app).
+
+## Authentication
+For more details about the authentication used in this integration, see [Microsoft Integrations - Authentication](https://xsoar.pan.dev/docs/reference/articles/microsoft-integrations---authentication).
```
Comment (JasBeilin) [inline Packs/AzureResourceGraph/Integrations/AzureResourceGraph/README.md:8]:
```suggestion
There are two application authentication methods available:
Follow your preferred method's guide on how to use the admin consent flow in order to receive your authentication information:

* [Cortex XSOAR Application](https://xsoar.pan.dev/docs/reference/articles/microsoft-integrations---authentication#cortex-xsoar-application)
* [Self-Deployed Application](https://xsoar.pan.dev/docs/reference/articles/microsoft-integrations---authentication#self-deployed-application)

For more details about the authentication used in this integration, see [Microsoft Integrations - Authentication](https://xsoar.pan.dev/docs/reference/articles/microsoft-integrations---authentication).

```

File: Packs/AzureResourceGraph/README.md [hunk 20]
```diff
@@ -0,0 +1,23 @@
...
+- [Link to starter Azure Resource Graph Queries](https://learn.microsoft.com/en-us/azure/governance/resource-graph/samples/starter?tabs=azure-cli)
+- [Link to advanced Azure Resource Graph Queries](https://learn.microsoft.com/en-us/azure/governance/resource-graph/samples/advanced?tabs=azure-cli)
+
+## Note from Microsoft
+
+> Throttling
+>
+> As a free service, queries to Resource Graph are throttled to provide the best experience and response time for all customers. If your organization wants to use the Resource Graph API for large-scale and frequent queries, use portal Feedback from the [Resource Graph portal page](https://portal.azure.com/#blade/HubsExtension/ArgQueryBlade).
+>
+> For more information, see [Guidance for throttled requests](https://learn.microsoft.com/en-us/azure/governance/resource-graph/concepts/guidance-for-throttled-requests).
```
Comment (JasBeilin) [inline Packs/AzureResourceGraph/README.md:23]:
```suggestion
```
Why is this needed?

**Review by JasBeilin — CHANGES_REQUESTED** [review]

File: Packs/AzureResourceGraph/README.md (hunk 20 above)
Comment (BigEasyJ) [reply Packs/AzureResourceGraph/README.md:23]: For customers to know that since we are using Azure's API that they have to make a change in Azure to fix any issues where throttling happens. I could see this being a question brought to the attention to CS in later scenarios based on experience.

File: Packs/AzureResourceGraph/Integrations/AzureResourceGraph/AzureResourceGraph.yml [hunk 21]
```diff
@@ -0,0 +1,142 @@
...
+      name: page
+    - name: management_groups
+      description: "Azure management groups against which to execute the query. Example: 'mg1, mg2'"
+    - name: subscriptions
+      description: "Azure subscriptions against which to execute the query. Example: 'sub1, sub2'"
+    - name: query
+      required: true
+      description: 'The query for the Azure Resource Graph to execute. The query is based on Kusto Query Language (KQL). Reference: https://learn.microsoft.com/en-us/azure/data-explorer/kusto/query/'
+    outputs:
+    - contextPath: AzureResourceGraph.Query
```
[the remark this answers is not in the document]
Comment (BigEasyJ) [reply Packs/AzureResourceGraph/Integrations/AzureResourceGraph/AzureResourceGraph.yml:131]: They are numerous amounts also there is the ability for custom fields.

**Review by JasBeilin — APPROVED** [review]
Thank you for your contribution! we will take it and continue to push it to production. We will contact you for every change if required.
